File: backend/arrlink/arr/radarr.py
"""Radarr adapter (v3 API)."""
from __future__ import annotations

import os

from .base import (
    AdapterError,
    AppInfo,
    BaseAdapter,
    Item,
    MediaFile,
    Tag,
    translate_tag_labels,
)


class RadarrAdapter(BaseAdapter):
    app_type = "radarr"

    async def ping(self) -> AppInfo:
        data = await self._get_json("/api/v3/system/status")
        if not isinstance(data, dict) or "version" not in data:
            raise AdapterError("unexpected system/status payload")
        return AppInfo(name="radarr", version=str(data["version"]))

    async def fetch_tags(self) -> list[Tag]:
        data = await self._get_json("/api/v3/tag")
        if not isinstance(data, list):
            raise AdapterError("unexpected tag payload")
        tags: list[Tag] = []
        for row in data:
            if not isinstance(row, dict):
                continue
            label = (row.get("label") or "").strip()
            if not label:
                continue
            try:
                count = int(row.get("count") or 0)
            except (TypeError, ValueError):
                count = 0
            try:
                tid = int(row.get("id"))
            except (TypeError, ValueError):
                tid = None
            tags.append(Tag(label=label, count=count, id=tid))
        return tags

    async def create_tag(self, label: str) -> None:
        import httpx

        label = (label or "").strip()
        if not label:
            raise AdapterError("tag label is empty")
        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                r = await client.post(
                    f"{self.url}/api/v3/tag",
                    json={"label": label},
                    headers={"X-Api-Key": self.api_key},
                )
        except Exception as e:  # noqa: BLE001
            raise AdapterError(f"unreachable: {e}") from e
        if r.status_code == 401 or r.status_code == 403:
            raise AdapterError("bad API key (401)", status=401)
        if r.status_code not in (200, 201):
            raise AdapterError(f"HTTP {r.status_code} creating tag '{label}'",
                              status=r.status_code)

    async def fetch_items(self, known_fingerprints=None) -> list[Item]:
        # Radarr's item list is a single /movie call, so there's nothing to
        # skip -- known_fingerprints is accepted for interface parity and
        # ignored.
        vocabulary = await self.fetch_tags()
        # Quality profile names aren't on the movie payload itself (only
        # qualityProfileId) — resolve via one extra call, same id->label
        # translation shape as the tag vocabulary above.
        try:
            profiles = await self.fetch_quality_profiles()
            profile_by_id = {p.id: p.name for p in profiles}
        except AdapterError:
            profile_by_id = {}
        data = await self._get_json("/api/v3/movie")
        if not isinstance(data, list):
            raise AdapterError("unexpected movie payload")
        items: list[Item] = []
        for row in data:
            if not isinstance(row, dict):
                continue
            movie_file = row.get("movieFile") or {}
            path = movie_file.get("path") if isinstance(movie_file, dict) else None
            if not path:
                continue  # not on disk yet
            labels = translate_tag_labels(vocabulary, row.get("tags"))
            try:
                year = row.get("year")
                year = int(year) if year else None
            except (TypeError, ValueError):
                year = None
            genres = [str(g).strip() for g in (row.get("genres") or []) if str(g).strip()]
            certification = (row.get("certification") or "").strip() or None
            collection_obj = row.get("collection")
            collection = (
                (collection_obj.get("name") or "").strip() or None
                if isinstance(collection_obj, dict)
                else None
            )
            qp_id = row.get("qualityProfileId")
            try:
                qp_id = int(qp_id) if qp_id is not None else None
            except (TypeError, ValueError):
                qp_id = None
            qp_name = profile_by_id.get(qp_id) if qp_id is not None else None
            lang_obj = row.get("originalLanguage")
            original_language = (
                (lang_obj.get("name") or "").strip() or None
                if isinstance(lang_obj, dict)
                else None
            )
            size = movie_file.get("size")
            try:
                size = int(size) if size is not None else None
            except (TypeError, ValueError):
                size = None
            item_dir = os.path.dirname(path)
            # stat the file so the poller can track inodes (Radarr's API
            # doesn't expose them); the container sees the same paths.
            try:
                st = os.stat(path)
                fsize = st.st_size
                fmtime = st.st_mtime
                finode = st.st_ino
            except OSError:
                fsize, fmtime, finode = size, None, None
            items.append(
                Item(
                    id=int(row["id"]),
                    title=str(row.get("title") or ""),
                    year=year,
                    tags=labels,
                    path=item_dir,
                    files=[
                        MediaFile(
                            rel_path=os.path.relpath(path, item_dir or "/"),
                            abs_path=str(path),
                            size=fsize,
                            mtime=fmtime,
                            inode=finode,
                        )
                    ],
                    genres=genres,
                    certification=certification,
                    collection=collection,
                    quality_profile_id=qp_id,
                    quality_profile_name=qp_name,
                    original_language=original_language,
                )
            )
        return items

File: backend-ts/src/arr/radarr.ts
import { statSync } from 'node:fs'
import { dirname, relative } from 'node:path'
import { BaseAdapter } from './base.js'
import {
  AdapterError,
  mediaInfoValues,
  toNumberOrNull,
  translateTagLabels,
  type AppInfo,
  type Item,
  type MediaFile,
  type MediaInfo,
  type Tag,
} from './types.js'

interface RadarrMovieFile {
  path?: string
  size?: number
  languages?: Array<{ id?: number; name?: string }>
  mediaInfo?: MediaInfo | null
}

interface RadarrMovieRow {
  id: number
  title?: string
  year?: number
  tags?: unknown
  movieFile?: RadarrMovieFile | null
  genres?: unknown
  certification?: string
  collection?: { title?: string; name?: string } | null
  qualityProfileId?: number | null
  originalLanguage?: { name?: string } | null
  studio?: string
  ratings?: {
    imdb?: { value?: number } | null
    tmdb?: { value?: number } | null
  } | null
  popularity?: number
  runtime?: number
}

export class RadarrAdapter extends BaseAdapter {
  readonly appType = 'radarr' as const

  protected itemPath(itemId: number): string {
    return `/api/v3/movie/${itemId}`
  }

  async ping(): Promise<AppInfo> {
    const data = (await this.getJson('/api/v3/system/status')) as { version?: unknown }
    if (!data || typeof data !== 'object' || !('version' in data)) {
      throw new AdapterError('unexpected system/status payload')
    }
    return { name: 'radarr', version: String(data.version) }
  }

  async fetchItems(
    _knownFingerprints?: Map<number, string>,
    tags?: Tag[],
  ): Promise<Item[]> {
    // known_fingerprints is accepted for interface parity and ignored:
    // Radarr's item list is a single /movie call.
    const [data, profileById, vocabulary] = await Promise.all([
      this.getJson('/api/v3/movie'),
      this.qualityProfileNames(),
      tags ? Promise.resolve(tags) : this.fetchTags(),
    ])
    if (!Array.isArray(data)) throw new AdapterError('unexpected movie payload')

    const items: Item[] = []
    for (const row of data as RadarrMovieRow[]) {
      if (!row || typeof row !== 'object') continue
      const movieFile = row.movieFile ?? {}
      const path = movieFile.path
      if (!path) continue // not on disk yet

      const labels = translateTagLabels(vocabulary, row.tags)
      const year = toNumberOrNull(row.year)
      const genres = (Array.isArray(row.genres) ? row.genres : [])
        .map((g) => String(g).trim())
        .filter(Boolean)
      const certification = (row.certification ?? '').trim() || null
      const collection = (row.collection?.title ?? row.collection?.name)?.trim() || null
      const qpId = toNumberOrNull(row.qualityProfileId)
      const qpName = qpId !== null ? (profileById.get(qpId) ?? null) : null
      const originalLanguage = row.originalLanguage?.name?.trim() || null
      // movieFile.languages is the file's own audio track(s), always present
      // alongside movieFile.path/size in the same /movie payload -- no extra call.
      const audioLanguages = (
        Array.isArray(movieFile.languages) ? movieFile.languages : []
      )
        .map((l) => l?.name?.trim())
        .filter((n): n is string => Boolean(n))
      const studio = (row.studio ?? '').trim() || null
      const media = mediaInfoValues(movieFile.mediaInfo)
      const rating = toNumberOrNull(row.ratings?.imdb?.value ?? row.ratings?.tmdb?.value)
      const popularity = toNumberOrNull(row.popularity)
      const runtime = toNumberOrNull(row.runtime)
      const apiSize = toNumberOrNull(movieFile.size)

      const itemDir = dirname(path)
      // Stat the file so the poller can track inodes (Radarr's API doesn't
      // expose them); the container sees the same paths.
      let fsize = apiSize
      let fmtime: number | null = null
      let finode: number | null = null
      try {
        const st = statSync(path)
        fsize = st.size
        fmtime = st.mtimeMs / 1000
        finode = st.ino
      } catch {
        // file not visible from this process -- keep the API-reported size
      }

      const file: MediaFile = {
        relPath: relative(itemDir || '/', path),
        absPath: String(path),
        size: fsize,
        mtime: fmtime,
        inode: finode,
      }

      items.push({
        id: Number(row.id),
        title: String(row.title ?? ''),
        year,
        tags: labels,
        path: itemDir,
        files: [file],
        genres,
        certification,
        collection,
        qualityProfileId: qpId,
        qualityProfileName: qpName,
        originalLanguage,
        audioLanguages,
        studio,
        ...media,
        rating,
        popularity,
        runtime,
      })
    }
    return items
  }
}
